scripts/observer.py

```diff
@@ -1,14 +1,260 @@
+if configs:
+  tracers = []
+
+  from opentelemetry import trace
+  from opentelemetry.sdk.resources import Resource
+  from opentelemetry.sdk.trace import TracerProvider
+  from opentelemetry.sdk.trace.export import BatchSpanProcessor
+  from opentelemetry.exporter.zipkin.json import ZipkinExporter
+  from opentelemetry.baggage.propagation import W3CBaggagePropagator
+  from opentelemetry.trace.propagation.tracecontext import TraceContextTextMapPropagator
+
+  import configparser
+  import os
+
+  def kvs_of_config(path,header="default"):
+    cfg = configparser.ConfigParser()
+    with open(path) as config_file:
+      cfg.read_string(f"[{header}]\n" + config_file.read())
+    kvs = dict(cfg[header])
+    kvs = {k: v.strip("'") for k, v in kvs.items()}
+    debug(path+":kvs="+str(kvs))
+    return kvs
+
+  try:
+    kvs_all_conf=kvs_of_config(observer_config_dir+"/all.conf")
+  except FileNotFoundError:
+    kvs_all_conf={}
+  module_names=kvs_all_conf.get("module_names", "LVHDSR,XenAPI,SR,SRCommand,util").split(",")
+  debug("module_names="+str(module_names))
+
+  def tracer_of_config(path):
+    otelvars='https://opentelemetry-python.readthedocs.io/en/latest/sdk/environment_variables.html'
+    argkv=kvs_of_config(path,header=otelvars)
+    config_otel_resource_attributes = argkv.get("otel_resource_attributes", "")
+    # Temporary: service.name is required and xapi is failing to write default attributes to the observer.conf
+    if "service.name" not in config_otel_resource_attributes:
+        config_otel_resource_attributes = "service.name=sm"
+    if config_otel_resource_attributes:
+      # OTEL requires some attributes e.g. service.name to be in the environment variable
+      os.environ["OTEL_RESOURCE_ATTRIBUTES"] = config_otel_resource_attributes
+    trace_log_dir = argkv.get("xs_exporter_bugtool_endpoint", "")
+    otel_exporter_zipkin_endpoints = argkv.get("xs_exporter_zipkin_endpoints").split(",") if argkv.get("xs_exporter_zipkin_endpoints") else []
+    otel_resource_attributes = dict(item.split("=") for item in argkv.get("otel_resource_attributes", "").split(",") if "=" in item)
+    service_name=argkv.get("otel_service_name", otel_resource_attributes.get("service.name", "unknown"))
+
+    host_uuid=otel_resource_attributes.get("xs.host.uuid", "unknown")
+    # Remove . to prevent users changing directories in the bugtool_filenamer
+    tracestate=os.getenv("TRACESTATE", "unknown").strip("'").replace(".", "")
+
+    from typing import Sequence
+    from opentelemetry.sdk.trace.export import SpanExportResult
+    from opentelemetry.trace import Span
+    from datetime import datetime, timezone
+    # eg.:"/var/log/dt/zipkinv2/json/xapi-6ddf2ff7-cdf7-479d-a943-ad8776c3fcf7-2023-11-09T18:39:53.322802-00:00.ndjson"
+    bugtool_filenamer = lambda: trace_log_dir + "/" + service_name + "-" + host_uuid + "-" + tracestate + "-" + datetime.now(timezone.utc).isoformat() + ".ndjson" # rfc3339
```

Comment:
This overlong string lambda would be much nicer to read and understand this way:
```suggestion
    def bugtool_filenamer()
        """Compound function to return an rfc3339-compilant ndjson file name"""
        now = datetime.now(timezone.utc).isoformat()
        f"{trace_log_dir}/{service_name}-{host_uuid}-{tracestate}-{now}.ndjson"
```